npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -1,21 +1,193 @@
 package ipsets
 
 import (
+	"bytes"
 	"fmt"
+	"os"
+	"strings"
 
-	"github.com/Azure/azure-container-networking/npm/util/errors"
-	"k8s.io/klog"
+	"github.com/Azure/azure-container-networking/npm/util"
+	kexec "k8s.io/utils/exec"
 )
 
+// TODO look at current errors in kusto
+
+// TODO eventually, have multiple retries and start at spot based on line number in error?
+const maxRetryCount = 1
+
+type ipsetRestoreFile struct {
+	lines      []string
+	retryCount int
+}
+
+var linuxExec kexec.Interface
+
+func assertExecExists() {
+	if linuxExec == nil {
+		linuxExec = kexec.New()
+	}
+}
+
+// don't need networkID
 func (iMgr *IPSetManager) applyIPSets(networkID string) error {
-	for setName := range iMgr.dirtyCaches {
-		set, exists := iMgr.setMap[setName] // check if the Set exists
-		if !exists {
-			return errors.Errorf(errors.AppendIPSet, false, fmt.Sprintf("member ipset %s does not exist", setName))
+	// DEBUGGING)
+	fmt.Println("DIRTY CACHE")
+	fmt.Println(iMgr.dirtyCaches)
+	fmt.Println("DELETE CACHE")
+	fmt.Println(iMgr.deleteCache)
+
+	file := initIPSetRestoreFile()
+	file.create(iMgr)
+
+	// MORE DEBUGGING
+	fmt.Println("RESTORE FILE")
+	fmt.Println(file.toString(0))
+
+	assertExecExists() // TODO remove if using osexec
+	for {
+		err := file.restore(0) // TODO could retry from line that fails
+		if err == nil {
+			return nil
+		}
+		file.retryCount += 1
+		if file.retryCount >= maxRetryCount {
+			return fmt.Errorf("failed to restore ipets after %d tries with final error: %w", maxRetryCount, err)
 		}
+	}
+}
+
+func initIPSetRestoreFile() *ipsetRestoreFile {
+	return &ipsetRestoreFile{
+		lines:      make([]string, 0),
+		retryCount: 0,
+	}
+}
+
+func (file *ipsetRestoreFile) create(iMgr *IPSetManager) {
+	// need to create all sets before possibly referencing them in lists
+	for setName := range iMgr.dirtyCaches {
+		set := iMgr.setMap[setName]
+		file.createSet(set)
+		// TEMP NOTES: uses ipset create --exist
+		//     alternatively, could maintain a create cache, but this complicates the dirty and delete cache
+	}
+
+	for setName := range iMgr.dirtyCaches {
+		set := iMgr.setMap[setName]
+		file.updateMembers(set)
+	}
+
+	for _, hashedName := range iMgr.deleteCache {
+		file.delete(hashedName)
+	}
+}
+
+func (file *ipsetRestoreFile) createSet(set *IPSet) {
+	methodFlag := util.IpsetNetHashFlag
+	if set.Kind == ListSet {
+		methodFlag = util.IpsetSetListFlag
+	} else if set.Type == NamedPorts {
+		methodFlag = util.IpsetIPPortHashFlag
+	}
+
+	specs := []string{util.IpsetCreationFlag, set.HashedName, util.IpsetExistFlag, methodFlag}
+	if set.Type == CIDRBlocks {
+		specs = append(specs, util.IpsetMaxelemName, util.IpsetMaxelemNum)
+	}
+
+	file.addLine(specs...)
+}
 
-		klog.Infof(set.Name)
+func (file *ipsetRestoreFile) updateMembers(set *IPSet) {
+	file.flush(set.HashedName)
 
+	// DEBUGGING
+	fmt.Printf("DEBUG-ME\nname: %s\nkind: %s\npodip: \n", set.Name, set.Kind)
+	fmt.Println(set.IPPodKey)
+	fmt.Println("members: ")
```

Comment:
How are we managing deleting existing members of an ipset ?

Reply:
in updateMembers(), we flush the set before adding each member